Content.Server/EntityEffects/Effects/ActivateRandom.cs

```diff
@@ -0,0 +1,166 @@
+
+        for (int i = 0; i < Cases.Length; i++)
+        {
+            var _case = Cases[i];
+            foreach (var activate in _case.Activates)
+            {
+                if (args.Source.TryGetReagent(activate.Reagent, out var quantity))
+                {
+                    var impact = quantity.Quantity.Float() * activate.Factor.Float();
+                    chances[i] += impact;
+                    chancesSum += impact;
+                }
+            }
+        }
+
+        // Next, select a "random" set. Sets of higher propotions are prioritized.
+        int index = 0;
+        IRobustRandom _random = IoCManager.Resolve<IRobustRandom>();
+        {
+            float randval = _random.NextFloat(0f, chancesSum);
+            for (int i = 0; i < Cases.Length && randval > 0f; i++) {
+                index = i;
+                randval -= chances[i];
+            }
+        }
+
+        // When some set is activated, ActivateRandom removes reagents of the set marked as "consumed".
+        // When the effects of the set are run, the amount of *consumed* reagents as passed to them.
+        // This is useful, for example, in conjuction with HealthChange's scaleByQuantity
+        FixedPoint2 consumed = 0f;
+        foreach (var activate in Cases[index].Activates) {
+            if (activate.Consume && args.Source.TryGetReagent(activate.Reagent, out var quantity)) {
+                consumed += quantity.Quantity;
+                args.Source.RemoveReagent(quantity);
+            }
+        }
+
+        // Update the quantity to the amount consumed, set reagent to null.
+        EntityEffectReagentArgs newArgs = new EntityEffectReagentArgs(
+            args.TargetEntity,
+            args.EntityManager,
+            args.OrganEntity,
+            args.Source,
+            consumed, // quantity, set to the amount of reagent consumed. Therefore, catalysts only affect the chance of the
+            // outcome happening, not how powerful it will be.
+            null, // reagent, set to null because it's not obvious which reagent to pass when multiple reagents are consumed
+            args.Method,
+            args.Scale);
+        foreach (var effect in Cases[index].Effects) {
+            if (!effect.ShouldApply(newArgs, _random))
+                continue;
+
+            if (effect.ShouldLog)
+            {
+                var _adminLogger = IoCManager.Resolve<ISharedAdminLogManager>();
+                _adminLogger.Add(
+                    LogType.ReagentEffect,
+                    effect.LogImpact,
+                    $"Metabolism effect {effect.GetType().Name:effect}"
+                    + $" of reagent {args.Reagent?.LocalizedName:reagent}"
+                    + $" applied on entity {args.TargetEntity:entity}"
+                    + $" at {args.EntityManager.GetComponent<TransformComponent>(args.TargetEntity).Coordinates:coordinates}"
+                );
+            }
+            effect.Effect(newArgs);
```

Comment:
As above:
1. newline
2. Don't resolve logger inside here.
3. Styleguide for no prefix _.